File: 2021_04_18_make_a_lisp_in_csharp/Mal/Mal.Tests/EvalTests.cs

using System;
using System.Linq;
using Microsoft.VisualStudio.TestTools.UnitTesting;
using static Mal.Types;
using static Mal.Reader;
using static Mal.EnvM;
using static Mal.Core;
using static PowerFP.LListM;
using static Mal.Program;
using PowerFP;

namespace Mal.Tests
{
    [TestClass]
    public class EvalTests
    {
        // [TestMethod]
        // public void EvalFuncTest()
        // {
        //     //var mal = Reader.ReadText("(+ 5 (* 2 3))");
        //     //var mal = Reader.ReadText("(* 2 3)");
        //     var mal = Reader.ReadText("{\"a\" (+ 7 8)}");
        //     var aa = Evaluation.Eval(mal!, Evaluation.Env);
        //     Console.WriteLine(mal.PrintStr(true));
        //     Console.WriteLine(aa.PrintStr(true));
        // }

        [TestMethod]
        public void ApplyDefTest()
        {
            var env1 = EmptyEnv();

            Assert.ThrowsException<Exception>(() => EvalM.ApplyDef(MalLListFrom(), env1));
            Assert.ThrowsException<Exception>(() => EvalM.ApplyDef(MalLListFrom(new Symbol("a", NilV)), env1));
            Assert.ThrowsException<Exception>(() => EvalM.ApplyDef(MalLListFrom(
                new Symbol("a", NilV), new Number(1), new Number(1)), env1));

            var mal1 = EvalM.ApplyDef(MalLListFrom(new Symbol("x", NilV), new Number(1)), env1);

            Assert.AreEqual(new Number(1), mal1);
            Assert.AreEqual(new Number(1), env1.Get(new Symbol("x", NilV)));


            var env2 = new Env(MapM.Empty<Symbol, MalType>(), DefaultEnv());
            var mal2 = EvalM.ApplyDef(MalLListFrom(
                new Symbol("y", NilV),
                MalListFrom(new Symbol("+", NilV), new Number(1), new Number(2))
                ), env2);

            Assert.AreEqual(new Number(3), mal2);
            Assert.AreEqual(new Number(3), env2.Get(new Symbol("y", NilV)));
        }

        [TestMethod]
        public void ApplyBindingTest()
        {
            Assert.ThrowsException<Exception>(() => EvalM.ApplyBindings(MalLListFrom(new Symbol("a", NilV)), EmptyEnv()));

            Assert.ThrowsException<Exception>(() => EvalM.ApplyBindings(
                MalLListFrom(new Symbol("a", NilV), new Number(1), new Symbol("b", NilV)), EmptyEnv()));

            var env1 = EmptyEnv();
            var envResult1 = EvalM.ApplyBindings(null, env1);
            Assert.AreSame(env1, envResult1);

            var env2 = EmptyEnv();
            var envResult2 = EvalM.ApplyBindings(MalLListFrom(
                new Symbol("a", NilV),
                new Number(1),
                new Symbol("b", NilV),
                new Number(2)
                ), env2);

            Assert.AreSame(env2, envResult2);
            Assert.AreEqual(new Number(1), env2.Get(new Symbol("a", NilV)));
            Assert.AreEqual(new Number(2), env2.Get(new Symbol("b", NilV)));
        }



        [TestMethod]
        public void ApplyLetTest()
        {
            Assert.ThrowsException<Exception>(() => EvalM.ApplyLet(MalLListFrom(new Symbol("a", NilV)), EmptyEnv()));
            Assert.ThrowsException<Exception>(() => EvalM.ApplyLet(MalLListFrom(MalListFrom(new Symbol("a", NilV), new Number(1))), EmptyEnv()));

            var malResult = EvalM.ApplyLet(MalLListFrom(
                MalListFrom(new Symbol("a", NilV), new Number(1), new Symbol("b", NilV), new Number(3)),
                MalListFrom(new Symbol("+", NilV), new Symbol("a", NilV), new Symbol("b", NilV))
                ), DefaultEnv());

            Assert.AreEqual(new Number(4), malResult);
        }


        [TestMethod]
        public void ApplyDoTest()
        {
            Assert.ThrowsException<Exception>(() => EvalM.ApplyDo(MalLListFrom(), EmptyEnv()));
            var malResult = EvalM.ApplyDo(MalLListFrom(NilV, MalListFrom(new Symbol("+", NilV), new Number(1), new Number(3))), DefaultEnv());
            Assert.AreEqual(new Number(4), malResult);
        }

        [TestMethod]
        public void ApplyIfTest()
        {
            Assert.ThrowsException<Exception>(() => EvalM.ApplyIf(MalLListFrom(
                ), EmptyEnv()));

            Assert.ThrowsException<Exception>(() => EvalM.ApplyIf(MalLListFrom(
                new Symbol("a", NilV)
                ), EmptyEnv()));

            Assert.ThrowsException<Exception>(() => EvalM.ApplyIf(MalLListFrom(
                new Symbol("a", NilV),
                new Symbol("a", NilV),
                new Symbol("a", NilV),
                new Symbol("a", NilV)
                ), EmptyEnv()));

            // falsy values
            Assert.AreEqual(new Number(2),
                EvalM.ApplyIf(MalLListFrom(FalseV, new Number(1), new Number(2)), EmptyEnv()));
            Assert.AreEqual(new Number(2),
                EvalM.ApplyIf(MalLListFrom(NilV, new Number(1), new Number(2)), EmptyEnv()));
            Assert.AreEqual(NilV,
                EvalM.ApplyIf(MalLListFrom(NilV, new Number(1)), EmptyEnv()));

            //truthy values
            Assert.AreEqual(new Number(1),
                EvalM.ApplyIf(MalLListFrom(TrueV, new Number(1), new Number(2)), EmptyEnv()));
            Assert.AreEqual(new Number(1),
                EvalM.ApplyIf(MalLListFrom(new Str(""), new Number(1), new Number(2)), EmptyEnv()));
            Assert.AreEqual(new Number(1),
                EvalM.ApplyIf(MalLListFrom(new List(null, ListType.Vector, NilV), new Number(1), new Number(2)), EmptyEnv()));
        }

        [TestMethod]
        public void BindFunctionArgumentsTest()
        {
            Symbol a = new Symbol("a", NilV), b = new Symbol("b", NilV), amp = new Symbol("&", NilV);
            MalType one = new Number(1), two = new Number(2), three = new Number(3);

            Assert.AreEqual(LListFrom((a, one)), EvalM.BindFunctionArguments(MalLListFrom(a), LListFrom(one)));
            Assert.AreEqual(null, EvalM.BindFunctionArguments(MalLListFrom(), LListFrom(one)));
            Assert.AreEqual(LListFrom((a, one)), EvalM.BindFunctionArguments(MalLListFrom(a), LListFrom(one, two)));
            Assert.ThrowsException<Exception>(() => EvalM.BindFunctionArguments(MalLListFrom(a), MalLListFrom()));

            Assert.AreEqual(
                LListFrom((a, one), (b, new List(LListFrom(two, three), ListType.List, NilV))),
                EvalM.BindFunctionArguments(MalLListFrom(a, amp, b), LListFrom(one, two, three)));
            Assert.AreEqual(
                LListFrom((a, one), (b, new List(null, ListType.List, NilV))),
                EvalM.BindFunctionArguments(MalLListFrom(a, amp, b), LListFrom(one)));

            Assert.ThrowsException<Exception>(() => Assert.AreEqual(
                LListFrom((a, one), (b, new List(null, ListType.List, NilV))),
                EvalM.BindFunctionArguments(MalLListFrom(a, amp), LListFrom(one))));
        }

        [TestMethod]
        public void ApplyFnTest()
        {
            Assert.ThrowsException<Exception>(() => EvalM.ApplyFn(MalLListFrom(), EmptyEnv()));
            Assert.ThrowsException<Exception>(() => EvalM.ApplyFn(MalLListFrom(new Number(1), new Number(1)), EmptyEnv()));
            Assert.ThrowsException<Exception>(() => EvalM.ApplyFn(MalLListFrom(MalListFrom(), new Number(1), new Number(1)), EmptyEnv()));


            Symbol a = new Symbol("a", NilV), b = new Symbol("b", NilV), plus = new Symbol("+", NilV);
            MalType one = new Number(1), two = new Number(2);

            var fn = (Fn)EvalM.ApplyFn(MalLListFrom(MalListFrom(a, b), MalListFrom(plus, a, b)), DefaultEnv());

            Assert.AreEqual(new Number(3), fn.Value(LListFrom(one, two)));
        }

        [TestMethod]
        public void ApplyQuoteTest()
        {
            Assert.ThrowsException<Exception>(() => EvalM.ApplyQuote(MalLListFrom(), EmptyEnv()));
            Assert.ThrowsException<Exception>(() => EvalM.ApplyFn(MalLListFrom(new Number(1), new Number(1)), EmptyEnv()));

            Assert.AreEqual(new Number(1), EvalM.ApplyQuote(MalLListFrom(new Number(1)), EmptyEnv()));
        }

        [TestMethod]
        public void TransformQuasiquoteTest()
        {
            Assert.AreEqual(new Number(1), EvalM.TransformQuasiquote(new Number(1)));
            Assert.AreEqual(
                MalListFrom(new Symbol("quote", NilV), new Symbol("abc", NilV)),
                EvalM.TransformQuasiquote(new Symbol("abc", NilV)));

            Assert.AreEqual("(cons 1 (cons 2 ()))", Printer.PrintStr(EvalM.TransformQuasiquote(
                MalListFrom(new Number(1), new Number(2)))));


            var oneTwoList = MalListFrom(new Number(1), new Number(2));

            var mal = EvalM.TransformQuasiquote(
                MalListFrom(
                    new Number(1),
                    MalListFrom(new Symbol("unquote", NilV), oneTwoList),
                    new Number(4),
                    MalListFrom(new Symbol("splice-unquote", NilV), oneTwoList)
                    ));

            Assert.AreEqual("(cons 1 (cons (1 2) (cons 4 (concat (1 2) ()))))", Printer.PrintStr(mal));
        }

        [TestMethod]
        public void ApplyDefMacroTest()
        {
            var env1 = EmptyEnv();

            Assert.ThrowsException<Exception>(() => EvalM.ApplyDefMacro(MalLListFrom(), env1));
            Assert.ThrowsException<Exception>(() => EvalM.ApplyDefMacro(MalLListFrom(new Symbol("a", NilV)), env1));
            Assert.ThrowsException<Exception>(() => EvalM.ApplyDefMacro(MalLListFrom(new Symbol("a", NilV), new Number(1)), env1));

            var mal1 = EvalM.ApplyDefMacro(MalLListFrom(
                new Symbol("x", NilV),
                MalListFrom(new Symbol("fn*", NilV), MalListFrom(), NilV)
                ), env1);

            Assert.IsTrue(mal1 is Fn { IsMacro: true });
        }


        [TestMethod]
        public void IsMacroCallTest()
        {

            var env1 = EmptyEnv();
            env1.Set(new Symbol("number", NilV), new Number(1));
            env1.Set(new Symbol("function", NilV), new Fn(args => NilV, NilV, false));
            env1.Set(new Symbol("macro", NilV), new Fn(args => NilV, NilV, true));

            Assert.ThrowsException<Exception>(() => EvalM.IsMacroCall(MalListFrom(new Symbol("number__", NilV)), env1));

            Assert.IsFalse(EvalM.IsMacroCall(MalListFrom(new Symbol("number", NilV)), env1));
            Assert.IsFalse(EvalM.IsMacroCall(MalListFrom(new Symbol("function", NilV)), env1));
            Assert.IsTrue(EvalM.IsMacroCall(MalListFrom(new Symbol("macro", NilV)), env1));

        }


        [TestMethod]
        public void TryCatchTest()
        {
            var emptyEnv = EmptyEnv();

            Assert.ThrowsException<Exception>(() => EvalM.ApplyTryCatch(MalLListFrom(TrueV), emptyEnv));
            Assert.ThrowsException<Exception>(() => EvalM.ApplyTryCatch(MalLListFrom(TrueV, MalListFrom(new Symbol("catch*", NilV), new Symbol("err", NilV), TrueV, FalseV)), emptyEnv));

            Assert.AreEqual(TrueV, EvalM.ApplyTryCatch(
                MalLListFrom(TrueV, MalListFrom(new Symbol("catch*", NilV), new Symbol("err", NilV), FalseV)), emptyEnv));

            Assert.AreEqual(FalseV, EvalM.ApplyTryCatch(
                MalLListFrom(new Symbol("xx", NilV), MalListFrom(new Symbol("catch*", NilV), new Symbol("err", NilV), FalseV)), emptyEnv));

            var defaultEnv = DefaultEnv();

            Assert.AreEqual(new Number(666), EvalM.ApplyTryCatch(
                MalLListFrom(MalListFrom(new Symbol("throw", NilV), new Number(666)), MalListFrom(new Symbol("catch*", NilV), new Symbol("err", NilV), new Symbol("err", NilV))), defaultEnv));

        }
    }
}

File: 2021_10_21_make_a_lisp_in_fsharp/MiniMal/MiniMal.Tests/EvalTests.fs
namespace MiniMal.Tests


open Microsoft.VisualStudio.TestTools.UnitTesting
open Types
open Eval
open TestUtils
open Printer

[<TestClass>]
type EvalTests() =

    [<TestMethod>]
    member this.ApplyDefTest() =
        let env1 = emptyEnv None
        assertThrowsException (fun () -> applyDef [] env1)
        assertThrowsException (fun () -> applyDef [ Symbol("a") ] env1)
        assertThrowsException (fun () -> applyDef [ Symbol("a"); Number(1.); Number(1.) ] env1)
        let mal1 =
            applyDef [ Symbol("x"); Number(1.) ] env1
        assertMalEquals (Number(1.)) mal1
        assertMalEquals (Number(1.)) (env1.Get "x")


    [<TestMethod>]
    member this.ApplyBindingTest() =
        assertThrowsException (fun () -> applyBindings [ Symbol("a") ] (emptyEnv None))
        assertThrowsException (fun () -> applyBindings [ Symbol("a"); Number(1.); Symbol("b") ] (emptyEnv None))

        let env1 = emptyEnv None
        let envResult1 = applyBindings [] env1
        Assert.AreSame(env1, envResult1)

        let env2 = emptyEnv None
        let envResult2 =
            applyBindings [ Symbol("a"); Number(1.); Symbol("b"); Number(2.) ] env2
        Assert.AreSame(env2, envResult2)
        assertMalEquals (Number(1.)) (env2.Get "a")
        assertMalEquals (Number(2.)) (env2.Get "b")


    [<TestMethod>]
    member this.ApplyLetTest() =
        assertThrowsException (fun () -> applyLet [ Symbol("a") ] (emptyEnv None))
        assertThrowsException (fun () -> applyLet [ Symbol("a"); Number(1.) ] (emptyEnv None))

        let malResult =
            applyLet
                [ MalList([ Symbol("a"); Number(1.); Symbol("b"); Number(3.) ], List)
                  MalList([ Symbol("+"); Symbol("a"); Symbol("b") ], List) ]
                (defaultEnv None)

        assertMalEquals (Number(4.)) malResult


    [<TestMethod>]
    member this.ApplyDoTest() =
        assertThrowsException (fun () -> applyDo [] (emptyEnv None))
        let malResult =
            applyDo [ Nil; MalList([ Symbol("+"); Number(1.); Number(3.) ], List) ] (defaultEnv None)

        assertMalEquals (Number(4.)) malResult

    [<TestMethod>]
    member this.ApplyIfTest() =
        assertThrowsException (fun () -> applyIf [] (emptyEnv None))
        assertThrowsException (fun () -> applyIf [ Symbol("a") ] (emptyEnv None))
        assertThrowsException (fun () -> applyIf [ Symbol("a"); Symbol("a"); Symbol("a"); Symbol("a") ] (emptyEnv None))
        // falsy values
        assertMalEquals (Number(2.)) (applyIf [ False; Number(1.); Number(2.) ] (emptyEnv None))
        assertMalEquals (Number(2.)) (applyIf [ Nil; Number(1.); Number(2.) ] (emptyEnv None))
        assertMalEquals Nil (applyIf [ Nil; Number(1.) ] (emptyEnv None))
        // truthy values
        assertMalEquals (Number(1.)) (applyIf [ True; Number(1.); Number(2.) ] (emptyEnv None))
        assertMalEquals (Number(1.)) (applyIf [ Str(""); Number(1.); Number(2.) ] (emptyEnv None))
        assertMalEquals (Number(1.)) (applyIf [ MalList([], Vector); Number(1.); Number(2.) ] (emptyEnv None))

    [<TestMethod>]
    member this.BindFunctionArgumentsTest() =
        let one = Number(1.)
        let two = Number(2.)
        let three = Number(3.)
        let comparePair (name1: string, mal1) (name2, mal2) = name1 = name2 && malEquals mal1 mal2
        Assert.IsTrue(
            compareLists
                [ "a", one ]
                (bindFunctionArguments [ Symbol("a") ] [
                    one
                 ])
                comparePair
        )
        Assert.IsTrue(compareLists [] (bindFunctionArguments [] [ one ]) comparePair)
        Assert.IsTrue(
            compareLists
                [ "a", one ]
                (bindFunctionArguments [ Symbol("a") ] [
                    one
                    two
                 ])
                comparePair
        )
        assertThrowsException (fun () -> bindFunctionArguments [ Symbol("a") ] [])
        Assert.IsTrue(
            compareLists
                [ ("a", one); ("b", MalList([ two; three ], List)) ]
                (bindFunctionArguments [ Symbol("a"); Symbol("&"); Symbol("b") ] [
                    one
                    two
                    three
                 ])
                comparePair
        )
        Assert.IsTrue(
            compareLists
                [ ("a", one); ("b", MalList([], List)) ]
                (bindFunctionArguments [ Symbol("a"); Symbol("&"); Symbol("b") ] [
                    one
                 ])
                comparePair
        )
        assertThrowsException
            (fun () ->
                bindFunctionArguments [ Symbol("a"); Symbol("&") ] [
                    one
                ])

    [<TestMethod>]
    member this.ApplyFnTest() =
        assertThrowsException (fun () -> applyFn [] (emptyEnv None))
        assertThrowsException (fun () -> applyFn [ Number(1.); Number(2.) ] (emptyEnv None))
        assertThrowsException (fun () -> applyFn [ MalList([ Number(1.); Number(2.) ], List) ] (emptyEnv None))
        let a, b, plus = Symbol("a"), Symbol("b"), Symbol("+")
        let one, two = Number(1.), Number(2.)
        let fn =
            applyFn [ MalList([ a; b ], List); MalList([ plus; a; b ], List) ] (defaultEnv None)
        match fn with
        | Fn (func, false) -> assertMalEquals (Number(3.)) (func [ one; two ])
        | _ -> ()

    [<TestMethod>]
    member this.ApplyQuoteTest() =
        assertThrowsException (fun () -> applyQuote [] (emptyEnv None))
        assertThrowsException (fun () -> applyQuote [ Number(1.); Number(1.) ] (emptyEnv None))
        assertMalEquals (Number(1.)) (applyQuote [ Number(1.) ] (emptyEnv None))


    [<TestMethod>]
    member this.TransformQuasiquoteTest() =
        assertMalEquals (Number(1.)) (transformQuasiquote (Number(1.)))
        assertMalEquals (MalList([ Symbol("quote"); Symbol("abc") ], List)) (transformQuasiquote (Symbol("abc")))
        let mal1 =
            transformQuasiquote (MalList([ Number(1.); Number(2.) ], List))
        Assert.AreEqual("(cons 1 (cons 2 ()))", printStr mal1)
        let oneTwoList =
            MalList([ Number(1.); Number(2.) ], List)
        let mal2 =
            transformQuasiquote (
                MalList(
                    [ Number(1.)
                      MalList([ Symbol("unquote"); oneTwoList ], List)
                      Number(4.)
                      MalList([ Symbol("splice-unquote"); oneTwoList ], List) ],
                    List
                )
            )
        Assert.AreEqual("(cons 1 (cons (1 2) (cons 4 (concat (1 2) ()))))", printStr mal2)

    [<TestMethod>]
    member this.ApplyDefMacroTest() =
        let env1 = emptyEnv None
        assertThrowsException (fun () -> applyDefMacro [] env1)
        assertThrowsException (fun () -> applyDefMacro [ Symbol("a") ] env1)
        assertThrowsException (fun () -> applyDefMacro [ Symbol("a"); Number(1.) ] env1)

        let mal1 =
            applyDefMacro [ Symbol("x"); MalList([ Symbol("fn*"); MalList([], List); Nil ], List) ] env1
        Assert.IsTrue(
            match mal1 with
            | Fn (_, true) -> true
            | _ -> false
        )

    [<TestMethod>]
    member this.IsMacroCallTest() =
        let env1 = emptyEnv None
        env1.Set "number" (Number(1.)) |> ignore
        env1.Set "function" (Fn((fun _ -> Nil), false))
        |> ignore
        env1.Set "macro" (Fn((fun _ -> Nil), true))
        |> ignore

        assertThrowsException (fun () -> isMacroCall (MalList([ Symbol("number__") ], List)) env1)

        Assert.IsFalse(isMacroCall (MalList([ Symbol("number") ], List)) env1)
        Assert.IsFalse(isMacroCall (MalList([ Symbol("function") ], List)) env1)
        Assert.IsTrue(isMacroCall (MalList([ Symbol("macro") ], List)) env1)


//         [TestMethod]
//         public void TryCatchTest()
//         {
//             var emptyEnv = EmptyEnv();

//             Assert.ThrowsException<Exception>(() => EvalM.ApplyTryCatch(MalLListFrom(TrueV), emptyEnv));
//             Assert.ThrowsException<Exception>(() => EvalM.ApplyTryCatch(MalLListFrom(TrueV, MalListFrom(new Symbol("catch*"), new Symbol("err"), TrueV, FalseV)), emptyEnv));

//             Assert.AreEqual(TrueV, EvalM.ApplyTryCatch(
//                 MalLListFrom(TrueV, MalListFrom(new Symbol("catch*"), new Symbol("err"), FalseV)), emptyEnv));

//             Assert.AreEqual(FalseV, EvalM.ApplyTryCatch(
//                 MalLListFrom(new Symbol("xx"), MalListFrom(new Symbol("catch*"), new Symbol("err"), FalseV)), emptyEnv));

//             var defaultEnv = DefaultEnv();

//             Assert.AreEqual(new Number(666), EvalM.ApplyTryCatch(
//                 MalLListFrom(MalListFrom(new Symbol("throw"), new Number(666)), MalListFrom(new Symbol("catch*"), new Symbol("err"), new Symbol("err"))), defaultEnv));

//         }
